### ISSUE TITLE
Support for interactive buffers

### ISSUE THREAD

**Issue by HamedHemati** [ContinualAI/avalanche#1164]
Some limitations with the current version of the `ReplayPlugin` make it hard (or even impossible) to implement some online CL strategies. 

Currently, for every new experience, `ReplayBuffer` creates a `ReplayDataloader` which goes through both the experience's dataset and the buffer's dataset in parallel. 

Here are some examples where this type of concatenation can be problematic: 
- If the buffer size is 100 and the online experience size is 1, it always creates a data loader of size 100 because of the buffer's dataset size.
- If the online experience size is 5, and for each sample, it needs to get 10 buffer samples, this wouldn't be possible with the current replay buffer (this is the behavior in an existing online CL strategy)

Usually, in replay-based online CL strategies, the number of samples concatenated to every batch of data is small, and we don't need to have the full buffer dataset with it, therefore we need a sampling-based interactive type of buffer that for each experience, randomly retrieves data from the buffer samples and adds them to the experience batch before `train_exp`/`before_training_iteration`.

I'm suggesting having another buffer plugin in addition to the current `ReplayPlguin`, we can call it `InteractiveReplayBuffer` or anything else that would be more appropriate for online strategies.  It would also be a nice feature to have something like a (binary) buffer mask that indicates which samples are from the buffer in case the strategy needs to treat them differently. 

Let me know what you think or if you have other suggestions. I currently have a local version of a similar plugin which I can adapt for the master branch.

**Comment by AntonioCarta** [comment]
If I understand correctly, the problem here is that using pytorch dataloaders is inefficient in OCL because the dataloader initialization is quite expensive and you want to avoid if you only need to sample a single minibatch.

To solve this, we don't need a new buffer type, we need to change the data loading for OCL. For example, we could add a sample method to the buffer and use that instead, like you are suggesting.

Instead of adding a new plugin, I would prefer to add a new argument to the existing one to customize the data loader. We could have:
- `dataloader` which is the current behavior
- `online_sampling`, your proposal
- `auto`, which would be the default and sets the loading type depending on the benchmark type.

**Comment by HamedHemati** [comment]
I'm closing this issue because it was mostly addressed in PR #1217. For special buffer update cases, the decision was to have a strategy-specific buffer.
